File: newfrontend/src/api/whiteboard.ts
import apiClient from './client';

export interface WhiteboardElement {
  id: string;
  board_id: string;
  type: string;
  x: number;
  y: number;
  width: number;
  height: number;
  rotation: number;
  z_index: number;
  text?: string;
  fill?: string;
  text_color?: string;
  font_family?: string;
  font_size?: number;
}

export interface WhiteboardConnection {
  id: string;
  board_id: string;
  source_element_id: string;
  target_element_id: string;
  stroke?: string;
  stroke_width?: number;
  points?: string | null;
}

export interface WhiteboardResponse {
  id: string;
  project_id: string;
  elements: WhiteboardElement[];
  connections: WhiteboardConnection[];
}

export async function getOrCreateWhiteboard(projectId: string): Promise<WhiteboardResponse> {
  const { data } = await apiClient.get('/whiteboards', { params: { project_id: projectId } });
  return data;
}

export async function createElement(boardId: string, payload: Partial<WhiteboardElement>): Promise<WhiteboardElement> {
  const { data } = await apiClient.post(`/whiteboards/${boardId}/elements`, payload);
  return data;
}

export async function updateElement(elementId: string, payload: Partial<WhiteboardElement>): Promise<WhiteboardElement> {
  const { data } = await apiClient.patch(`/whiteboard-elements/${elementId}`, payload);
  return data;
}

export async function deleteElement(elementId: string): Promise<void> {
  await apiClient.delete(`/whiteboard-elements/${elementId}`);
}

export async function createConnection(boardId: string, payload: Partial<WhiteboardConnection>): Promise<WhiteboardConnection> {
  const { data } = await apiClient.post(`/whiteboards/${boardId}/connections`, payload);
  return data;
}

export async function updateConnection(connectionId: string, payload: Partial<WhiteboardConnection>): Promise<WhiteboardConnection> {
  const { data } = await apiClient.patch(`/whiteboard-connections/${connectionId}`, payload);
  return data;
}

export async function deleteConnection(connectionId: string): Promise<void> {
  await apiClient.delete(`/whiteboard-connections/${connectionId}`);
}

File: backend/app/models/whiteboard.py
from sqlalchemy import Column, String, Text, Integer, ForeignKey, DateTime
from sqlalchemy.orm import relationship
from ..database import Base
import uuid
from datetime import datetime


class Whiteboard(Base):
    __tablename__ = "whiteboards"

    id = Column(String, primary_key=True, index=True, default=lambda: str(uuid.uuid4()))
    project_id = Column(String, ForeignKey("projects.id", ondelete="CASCADE"), nullable=False, index=True)
    created_at = Column(DateTime, default=datetime.utcnow)
    updated_at = Column(DateTime, default=datetime.utcnow, onupdate=datetime.utcnow)

    project = relationship("Project")
    elements = relationship("WhiteboardElement", back_populates="board", cascade="all, delete-orphan")
    connections = relationship("WhiteboardConnection", back_populates="board", cascade="all, delete-orphan")


class WhiteboardElement(Base):
    __tablename__ = "whiteboard_elements"

    id = Column(String, primary_key=True, index=True, default=lambda: str(uuid.uuid4()))
    board_id = Column("whiteboard_id", String, ForeignKey("whiteboards.id", ondelete="CASCADE"), nullable=False, index=True)
    type = Column(String, nullable=False, default="sticky")
    x = Column(Integer, default=0)
    y = Column(Integer, default=0)
    width = Column(Integer, default=160)
    height = Column(Integer, default=120)
    rotation = Column(Integer, default=0)
    z_index = Column(Integer, default=0)
    text = Column(Text, nullable=True)
    fill = Column(String, nullable=True)
    created_at = Column(DateTime, default=datetime.utcnow)
    updated_at = Column(DateTime, default=datetime.utcnow, onupdate=datetime.utcnow)

    board = relationship("Whiteboard", back_populates="elements")


class WhiteboardConnection(Base):
    __tablename__ = "whiteboard_connections"

    id = Column(String, primary_key=True, index=True, default=lambda: str(uuid.uuid4()))
    board_id = Column("whiteboard_id", String, ForeignKey("whiteboards.id", ondelete="CASCADE"), nullable=False, index=True)
    source_element_id = Column(String, ForeignKey("whiteboard_elements.id", ondelete="CASCADE"), nullable=False)
    target_element_id = Column(String, ForeignKey("whiteboard_elements.id", ondelete="CASCADE"), nullable=False)
    stroke = Column(String, default="#2b2d42")
    stroke_width = Column(Integer, default=2)
    points = Column(Text, nullable=True)  # JSON-serialized polyline points if needed
    created_at = Column(DateTime, default=datetime.utcnow)
    updated_at = Column(DateTime, default=datetime.utcnow, onupdate=datetime.utcnow)

    board = relationship("Whiteboard", back_populates="connections")


